pkgs/jni/lib/src/jclass.dart

```diff
@@ -0,0 +1,154 @@
+// Copyright (c) 2024, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+part of 'types.dart';
+
+/// A high level wrapper over a JNI class reference.
+///
+/// JClasses are objects of type `java.lang.Class`.
+class JClass extends JObject {
+  /// Constructs a [JClass] with the underlying [reference].
+  JClass.fromReference(JReference reference) : super.fromReference(reference);
+
+  /// Constructs a [JClass] associated with the class or interface with
+  /// the given string name.
+  JClass.forName(String name)
+      : super.fromReference(using((arena) {
+          final cls = Jni.accessors.getClass(name.toNativeChars(arena));
+          return JGlobalReference(cls.checkedClassRef);
+        }));
+
+  JConstructor constructor(String signature) {
+    return JConstructor._(this, signature);
+  }
+
+  JInstanceMethod instanceMethod(String name, String signature) {
+    return JInstanceMethod._(this, name, signature);
+  }
+
+  JStaticMethod staticMethod(String name, String signature) {
+    return JStaticMethod._(this, name, signature);
+  }
+
+  JInstanceField instanceField(String name, String signature) {
+    return JInstanceField._(this, name, signature);
+  }
+
+  JStaticField staticField(String name, String signature) {
+    return JStaticField._(this, name, signature);
+  }
+}
+
+/// A thin wrapper over a [JFieldIDPtr] of an instance field.
+extension type JInstanceField._fromPointer(JFieldIDPtr pointer) {
+  JInstanceField._(JClass jClass, String name, String signature)
+      : pointer = using((arena) => Jni.accessors
+            .getFieldID(
+              jClass.reference.pointer,
+              name.toNativeChars(arena),
+              signature.toNativeChars(arena),
+            )
+            .fieldID);
+
+  DartT get<JavaT, DartT>(JObject object, JAccessible<JavaT, DartT> type) {
+    return type._instanceGet(object.reference.pointer, this as JFieldIDPtr);
+  }
+
+  void set<JavaT, DartT>(
+      JObject object, JAccessible<JavaT, DartT> type, DartT value) {
+    type._instanceSet(object.reference.pointer, this as JFieldIDPtr, value);
+  }
+}
+
+/// A thin wrapper over a [JFieldIDPtr] of an static field.
+extension type JStaticField._fromPointer(JFieldIDPtr pointer) {
+  JStaticField._(JClass jClass, String name, String signature)
+      : pointer = using((arena) => Jni.accessors
+            .getStaticFieldID(
+              jClass.reference.pointer,
+              name.toNativeChars(arena),
+              signature.toNativeChars(arena),
+            )
+            .fieldID);
+
+  DartT get<JavaT, DartT>(JClass jClass, JAccessible<JavaT, DartT> type) {
```

Comment:
Consider making this an actual object so the `jClass` doesn't have to be passed again.